### ISSUE TITLE
Duplicate ports warning

### ISSUE THREAD

**Issue by sebhoss** [prometheus-operator/prometheus-operator#6284]
## Description

I recently updated prometheus-operator and saw the following line in the logs of the operator:

```
level=warn ts=2024-02-05T06:00:35.226955216Z caller=klog.go:106 component=k8s_client_runtime func=Warning msg="spec.template.spec.containers[1].ports[0]: duplicate port definition with spec.template.spec.initContainers[0].ports[0]"
```

I wasn't sure which pod actually declared duplicate ports, so after a short search I noticed that all prometheus pods now have both an initContainer (init-config-reloader) as well as a regular container (config-reloader) running on port 8080. Since I'm not the one who declared those ports/containers, I'm not sure whether I should do something about this warning.

## Steps to Reproduce

- Deploy latest prometheus-operator
- Deploy minimal `Prometheus` resource
- Watch logs of operator

## Expected Result

No warning should appear or at least the warning should tell me what to do.

## Actual Result

Warning confuses me & I want to go home

### Prometheus Operator Version

```shell
v0.71.2
```

### Kubernetes Version

```yaml
v1.28.6
```

### Kubernetes Cluster Type

kubeadm

### How did you deploy Prometheus-Operator?

yaml manifests

### Manifests

```yaml
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  labels:
    app.kubernetes.io/component: prometheus
    app.kubernetes.io/instance: autoscaling
    app.kubernetes.io/name: prometheus
    app.kubernetes.io/part-of: monitoring
  name: autoscaling
  namespace: prometheus
spec:
  additionalAlertManagerConfigs:
    key: prometheus-configuration
    name: alertmanager-configuration
    optional: false
  enableFeatures:
    - auto-gomaxprocs
  enableAdminAPI: true
  enforcedSampleLimit: 100
  evaluationInterval: 30s
  image: harbor.infra.run/dockerhub/prom/prometheus:v2.49.1
  imagePullSecrets:
    - name: harbor-infra-run-pull-secret
  nodeSelector:
    kubernetes.io/os: linux
  podMetadata:
    labels:
      app.kubernetes.io/component: prometheus
      app.kubernetes.io/instance: autoscaling
      app.kubernetes.io/name: prometheus
      app.kubernetes.io/part-of: monitoring
  podMonitorNamespaceSelector:
    matchLabels:
      autoscaling.prometheus.kube.infra.run/discovery: enabled
  podMonitorSelector:
    matchLabels:
      autoscaling.prometheus.kube.infra.run/discovery: enabled
  replicas: 2
  replicaExternalLabelName: prometheus_replica
  shards: 1
  retention: '6h'
  ruleNamespaceSelector:
    matchLabels:
      autoscaling.prometheus.kube.infra.run/discovery: enabled
  ruleSelector:
    matchLabels:
      autoscaling.prometheus.kube.infra.run/discovery: enabled
  scrapeInterval: 15s
  scrapeTimeout: 10s
  securityContext:
    fsGroup: 2000
    runAsNonRoot: true
    runAsUser: 1000
    runAsGroup: 2000
  serviceAccountName: prometheus-autoscaling
  serviceMonitorNamespaceSelector:
    matchLabels:
      autoscaling.prometheus.kube.infra.run/discovery: enabled
  serviceMonitorSelector:
    matchLabels:
      autoscaling.prometheus.kube.infra.run/discovery: enabled
  storage:
    volumeClaimTemplate:
      metadata:
        name: prometheus-autoscaling-data
        labels:
          app.kubernetes.io/component: prometheus
          app.kubernetes.io/instance: autoscaling
          app.kubernetes.io/name: prometheus
          app.kubernetes.io/part-of: monitoring
      spec:
        accessModes:
          - ReadWriteOnce
        storageClassName: csi-rbd
  thanos:
    image: harbor.infra.run/quay/thanos/thanos:v0.34.0
  web:
    pageTitle: Prometheus Autoscaling
```

### prometheus-operator log output

```shell
level=info ts=2024-02-05T06:00:34.810681Z caller=main.go:181 msg="Starting Prometheus Operator" version="(version=0.71.2, branch=refs/tags/v0.71.2, revision=af2014407bdc25c4fc2b26cd99c9655235ebdf88)"
level=info ts=2024-02-05T06:00:34.810751593Z caller=main.go:182 build_context="(go=go1.21.6, platform=linux/amd64, user=Action-Run-ID-7656327832, date=20240125-14:57:39, tags=unknown)"
level=info ts=2024-02-05T06:00:34.810767765Z caller=main.go:193 msg="namespaces filtering configuration " config="{allow_list=\"\",deny_list=\"\",prometheus_allow_list=\"\",alertmanager_allow_list=\"\",alertmanagerconfig_allow_list=\"\",thanosruler_allow_list=\"\"}"
level=info ts=2024-02-05T06:00:34.825852745Z caller=main.go:222 msg="connection established" cluster-version=v1.28.6
level=info ts=2024-02-05T06:00:34.855289664Z caller=operator.go:333 component=prometheus-controller msg="Kubernetes API capabilities" endpointslices=true
level=info ts=2024-02-05T06:00:34.874428002Z caller=operator.go:319 component=prometheusagent-controller msg="Kubernetes API capabilities" endpointslices=true
level=info ts=2024-02-05T06:00:34.893257626Z caller=server.go:298 msg="starting insecure server" address=[::]:8080
level=info ts=2024-02-05T06:00:34.992779203Z caller=operator.go:428 component=prometheusagent-controller msg="successfully synced all caches"
level=info ts=2024-02-05T06:00:34.994188604Z caller=operator.go:280 component=thanos-controller msg="successfully synced all caches"
level=info ts=2024-02-05T06:00:34.99522074Z caller=operator.go:542 component=thanos-controller key=prometheus/thanos-ruler-global msg="sync thanos-ruler"
level=info ts=2024-02-05T06:00:34.99545999Z caller=operator.go:311 component=alertmanager-controller msg="successfully synced all caches"
level=info ts=2024-02-05T06:00:34.999846973Z caller=operator.go:390 component=prometheus-controller msg="successfully synced all caches"
level=info ts=2024-02-05T06:00:35.003405968Z caller=operator.go:987 component=prometheus-controller key=prometheus/general msg="sync prometheus"
level=info ts=2024-02-05T06:00:35.164126655Z caller=operator.go:542 component=thanos-controller key=prometheus/thanos-ruler-global msg="sync thanos-ruler"
level=warn ts=2024-02-05T06:00:35.226955216Z caller=klog.go:106 component=k8s_client_runtime func=Warning msg="spec.template.spec.containers[1].ports[0]: duplicate port definition with spec.template.spec.initContainers[0].ports[0]"
level=warn ts=2024-02-05T06:00:35.309322232Z caller=klog.go:106 component=k8s_client_runtime func=Warning msg="spec.template.spec.containers[1].ports[0]: duplicate port definition with spec.template.spec.initContainers[0].ports[0]"
level=info ts=2024-02-05T06:00:35.319266257Z caller=operator.go:987 component=prometheus-controller key=prometheus/heavy msg="sync prometheus"
level=info ts=2024-02-05T06:00:35.344992244Z caller=operator.go:542 component=thanos-controller key=prometheus/thanos-ruler-global msg="sync thanos-ruler"
level=info ts=2024-02-05T06:00:35.428711029Z caller=operator.go:542 component=thanos-controller key=prometheus/thanos-ruler-global msg="sync thanos-ruler"
level=warn ts=2024-02-05T06:00:35.508559635Z caller=klog.go:106 component=k8s_client_runtime func=Warning msg="spec.template.spec.containers[1].ports[0]: duplicate port definition with spec.template.spec.initContainers[0].ports[0]"
level=info ts=2024-02-05T06:00:35.517673114Z caller=operator.go:987 component=prometheus-controller key=prometheus/autoscaling msg="sync prometheus"
level=warn ts=2024-02-05T06:00:35.677154596Z caller=klog.go:106 component=k8s_client_runtime func=Warning msg="spec.template.spec.containers[1].ports[0]: duplicate port definition with spec.template.spec.initContainers[0].ports[0]"
level=info ts=2024-02-05T06:00:35.685996793Z caller=operator.go:987 component=prometheus-controller key=prometheus/lernraum msg="sync prometheus"
level=warn ts=2024-02-05T06:00:35.862035321Z caller=klog.go:106 component=k8s_client_runtime func=Warning msg="spec.template.spec.containers[1].ports[0]: duplicate port definition with spec.template.spec.initContainers[0].ports[0]"
level=info ts=2024-02-05T06:00:35.866906713Z caller=operator.go:987 component=prometheus-controller key=prometheus/meta msg="sync prometheus"
level=warn ts=2024-02-05T06:00:36.048482343Z caller=klog.go:106 component=k8s_client_runtime func=Warning msg="spec.template.spec.containers[1].ports[0]: duplicate port definition with spec.template.spec.initContainers[0].ports[0]"
```

### Anything else?

I'm not sure whether the warning appeared in earlier versions..

**Comment by simonpasquier** [comment]
Thanks for the report! Indeed I can reproduce, it must be a change in Kubernetes v1.28 or v1.29 because I didn't see the warning with v1.27.

```
spec:
  template:
    spec:
      containers:
[...]
      - args:
        - --listen-address=:8080
        - --reload-url=http://localhost:9090/-/reload
        - --config-file=/etc/prometheus/config/prometheus.yaml.gz
        - --config-envsubst-file=/etc/prometheus/config_out/prometheus.env.yaml
        - --watched-dir=/etc/prometheus/rules/prometheus-prometheus-rulefiles-0
        command:
        - /bin/prometheus-config-reloader
        env:
        - name: POD_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.name
        - name: SHARD
          value: "0"
        image: quay.io/prometheus-operator/prometheus-config-reloader:v0.71.2
        imagePullPolicy: IfNotPresent
        name: config-reloader
        ports:
        - containerPort: 8080
          name: reloader-web
          protocol: TCP
[...]
      initContainers:
      - args:
        - --watch-interval=0
        - --listen-address=:8080
        - --config-file=/etc/prometheus/config/prometheus.yaml.gz
        - --config-envsubst-file=/etc/prometheus/config_out/prometheus.env.yaml
        - --watched-dir=/etc/prometheus/rules/prometheus-prometheus-rulefiles-0
        command:
        - /bin/prometheus-config-reloader
        env:
        - name: POD_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.name
        - name: SHARD
          value: "0"
        image: quay.io/prometheus-operator/prometheus-config-reloader:v0.71.2
        imagePullPolicy: IfNotPresent
        name: init-config-reloader
        ports:
        - containerPort: 8080
          name: reloader-web
          protocol: TCP
```

**Comment by sebhoss** [comment]
oh yeah you are right - that must be https://github.com/kubernetes/kubernetes/pull/113245 which was introduced in 1.28

**Comment by sebhoss** [comment]
reading the test cases in that PR it kinda looks like this warning is not supposed to happen because there is no test case that checks containers and initContainers

**Comment by simonpasquier** [comment]
Great finding! Let's see the answer from kube maintainers. Because of https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/, it might be "expected" that init and non-init containers don't share the same port number.

**Comment by tankbusta** [comment]
Hey folks 👋🏻 - I'm having an issue setting up an operator for the first time (on Kubernetes 1.29) where the Alert Manager & Prometheus StatefulSet end up in a perpetual termination loop. I've narrowed it down to something in the operator because scaling it to 0 will allow both the StatefulSet's to progress and everything comes up successfully.

I'm wondering if this could be the reason why as it's the **only** thing in the logs and it seems the Operator may be deleting the pods?

**Comment by simonpasquier** [comment]
@tankbusta I'm not sure how you problem relates to this issue. Please check https://prometheus-operator.dev/docs/operator/troubleshooting/#prometheusalertmanager-pods-stuck-in-terminating-loop-with-healthy-start-up-logs

**Comment by diranged** [comment]
@simonpasquier So this issue is definitely something new that cropped up in the [0.73.0](https://github.com/prometheus-operator/prometheus-operator/releases/tag/v0.73.0) release. If I run the operator in a Kind environment (1.26.6, and 1.29.2 both tested) I get a constant loop where the operator keeps killing and resyncing the AlertManagers and Thanos Rulers:

```
level=info ts=2024-04-06T00:01:17.533270048Z caller=operator.go:471 component=thanos-controller key=observability/default msg="sync thanos-ruler"
level=info ts=2024-04-06T00:01:17.614336673Z caller=operator.go:766 component=prometheus-controller key=observability/scraper msg="sync prometheus"
level=warn ts=2024-04-06T00:01:17.629025757Z caller=klog.go:106 component=k8s_client_runtime func=Warning msg="spec.template.spec.containers[1].ports[0]: duplicate port definition with spec.template.spec.initContainers[0].ports[0]"
level=info ts=2024-04-06T00:01:17.629379048Z caller=operator.go:572 component=alertmanager-controller key=observability/alertmanager msg="sync alertmanager"
level=info ts=2024-04-06T00:01:17.71029034Z caller=operator.go:471 component=thanos-controller key=observability/default msg="sync thanos-ruler"
level=warn ts=2024-04-06T00:01:17.738319007Z caller=klog.go:106 component=k8s_client_runtime func=Warning msg="spec.template.spec.containers[1].ports[0]: duplicate port definition with spec.template.spec.initContainers[0].ports[0]"
level=info ts=2024-04-06T00:01:17.741981132Z caller=operator.go:572 component=alertmanager-controller key=observability/alertmanager msg="sync alertmanager"
level=warn ts=2024-04-06T00:01:17.768118048Z caller=promcfg.go:1374 component=prometheus-controller msg="'bearerTokenFile' is deprecated, use 'authorization' instead." version=v2.51.1 service_monitor=observability/alertmanager-kube-promethe-coredns
level=warn ts=2024-04-06T00:01:17.768276923Z caller=promcfg.go:1374 component=prometheus-controller msg="'bearerTokenFile' is deprecated, use 'authorization' instead." version=v2.51.1 service_monitor=observability/prometheus-kube-prometheus-apiserver
level=warn ts=2024-04-06T00:01:17.768290298Z caller=promcfg.go:1374 component=prometheus-controller msg="'bearerTokenFile' is deprecated, use 'authorization' instead." version=v2.51.1 service_monitor=observability/prometheus-kube-prometheus-coredns
level=warn ts=2024-04-06T00:01:17.768301673Z caller=promcfg.go:1374 component=prometheus-controller msg="'bearerTokenFile' is deprecated, use 'authorization' instead." version=v2.51.1 service_monitor=observability/prometheus-kube-prometheus-kube-controller-manager
level=warn ts=2024-04-06T00:01:17.76831184Z caller=promcfg.go:1374 component=prometheus-controller msg="'bearerTokenFile' is deprecated, use 'authorization' instead." version=v2.51.1 service_monitor=observability/prometheus-kube-prometheus-kube-etcd
level=warn ts=2024-04-06T00:01:17.768322382Z caller=promcfg.go:1374 component=prometheus-controller msg="'bearerTokenFile' is deprecated, use 'authorization' instead." version=v2.51.1 service_monitor=observability/prometheus-kube-prometheus-kube-proxy
level=warn ts=2024-04-06T00:01:17.768333298Z caller=promcfg.go:1374 component=prometheus-controller msg="'bearerTokenFile' is deprecated, use 'authorization' instead." version=v2.51.1 service_monitor=observability/prometheus-kube-prometheus-kube-scheduler
level=warn ts=2024-04-06T00:01:17.826784257Z caller=klog.go:106 component=k8s_client_runtime func=Warning msg="spec.template.spec.containers[1].ports[0]: duplicate port definition with spec.template.spec.initContainers[0].ports[0]"
level=info ts=2024-04-06T00:01:17.831138965Z caller=operator.go:766 component=prometheus-controller key=observability/scraper msg="sync prometheus"
level=warn ts=2024-04-06T00:01:17.863252798Z caller=klog.go:106 component=k8s_client_runtime func=Warning msg="spec.template.spec.containers[1].ports[0]: duplicate port definition with spec.template.spec.initContainers[0].ports[0]"
level=info ts=2024-04-06T00:01:17.863601923Z caller=operator.go:572 component=alertmanager-controller key=observability/alertmanager msg="sync alertmanager"
level=info ts=2024-04-06T00:01:17.910792007Z caller=operator.go:471 component=thanos-controller key=observability/default msg="sync thanos-ruler"
level=warn ts=2024-04-06T00:01:18.008474923Z caller=klog.go:106 component=k8s_client_runtime func=Warning msg="spec.template.spec.containers[1].ports[0]: duplicate port definition with spec.template.spec.initContainers[0].ports[0]"
level=info ts=2024-04-06T00:01:18.009143465Z caller=operator.go:572 component=alertmanager-controller key=observability/alertmanager msg="sync alertmanager"
level=warn ts=2024-04-06T00:01:18.043482424Z caller=promcfg.go:1374 component=prometheus-controller msg="'bearerTokenFile' is deprecated, use 'authorization' instead." version=v2.51.1 service_monitor=observability/alertmanager-kube-promethe-coredns
level=warn ts=2024-04-06T00:01:18.043598215Z caller=promcfg.go:1374 component=prometheus-controller msg="'bearerTokenFile' is deprecated, use 'authorization' instead." version=v2.51.1 service_monitor=observability/prometheus-kube-prometheus-apiserver
level=warn ts=2024-04-06T00:01:18.04361659Z caller=promcfg.go:1374 component=prometheus-controller msg="'bearerTokenFile' is deprecated, use 'authorization' instead." version=v2.51.1 service_monitor=observability/prometheus-kube-prometheus-coredns
level=warn ts=2024-04-06T00:01:18.043627507Z caller=promcfg.go:1374 component=prometheus-controller msg="'bearerTokenFile' is deprecated, use 'authorization' instead." version=v2.51.1 service_monitor=observability/prometheus-kube-prometheus-kube-controller-manager
level=warn ts=2024-04-06T00:01:18.043645965Z caller=promcfg.go:1374 component=prometheus-controller msg="'bearerTokenFile' is deprecated, use 'authorization' instead." version=v2.51.1 service_monitor=observability/prometheus-kube-prometheus-kube-etcd
level=warn ts=2024-04-06T00:01:18.043658674Z caller=promcfg.go:1374 component=prometheus-controller msg="'bearerTokenFile' is deprecated, use 'authorization' instead." version=v2.51.1 service_monitor=observability/prometheus-kube-prometheus-kube-proxy
level=warn ts=2024-04-06T00:01:18.043670382Z caller=promcfg.go:1374 component=prometheus-controller msg="'bearerTokenFile' is deprecated, use 'authorization' instead." version=v2.51.1 service_monitor=observability/prometheus-kube-prometheus-kube-scheduler
level=warn ts=2024-04-06T00:01:18.158803007Z caller=klog.go:106 component=k8s_client_runtime func=Warning msg="spec.template.spec.containers[1].ports[0]: duplicate port definition with spec.template.spec.initContainers[0].ports[0]"
level=info ts=2024-04-06T00:01:18.15981009Z caller=operator.go:572 component=alertmanager-controller key=observability/alertmanager msg="sync alertmanager"
level=warn ts=2024-04-06T00:01:18.187363549Z caller=klog.go:106 component=k8s_client_runtime func=Warning msg="spec.template.spec.containers[1].ports[0]: duplicate port definition with spec.template.spec.initContainers[0].ports[0]"
level=info ts=2024-04-06T00:01:18.187736132Z caller=operator.go:766 component=prometheus-controller key=observability/scraper msg="sync prometheus"
```

If I back out to the [0.72.0](https://github.com/prometheus-operator/prometheus-operator/releases/tag/v0.72.0) release, it works fine. I've also tried building the images locally on my laptop and I get the same behavior ... building 0.72.0 works, but anything newer fails.

**Comment by simonpasquier** [comment]
@diranged can you open a new issue? I'm quite sure that the `duplicate port definition` log has nothing to do with objects being reconciled over and over.

**Comment by diranged** [comment]
> @diranged can you open a new issue? I'm quite sure that the `duplicate port definition` log has nothing to do with objects being reconciled over and over.

Done in https://github.com/prometheus-operator/prometheus-operator/issues/6503

**Comment by ausias-armesto** [comment]
I had a similar issue with the loop and the reason was that I had 2 instances of the prometheus operator running in different namespaces

**Comment by VtG242** [comment]
@ausias-armesto Did you solve it or you have install it only to one namespace?

**Comment by ausias-armesto** [comment]
To be honest, I don't know how it was installed in the `default` namespace, as I have everything using gitops practices, and the expected namespace was `monitoring`.
By removing the instance from the `default` namespace got solved

**Comment by VtG242** [comment]
I did a migration from kube-prometheus helmchart to kube-prometheus-stack helmchart (KPS ) and I wanted to have each installation in specific namespace and manage only  that namespace but It didn't work with recent version of KPS(v77.10.0) and I saw these errors. It worked with KPS version  v61.1.0.  At the end I have ended with only one KPS installation in one namespace and everything worked as expected.
